Priority multiplier

efim opened the pull request:
From this ticket (https://github.com/HabitRPG/habitrpg/issues/3556) and this trello page (https://trello.com/c/4C8w1z5h/17-task-difficulty-settings-v2-priority-multiplier)

Additional buttons for difficulty settings under the priority settings in the advanced options.

Commit:
copied part of jade that makes buttons for task difficulty

Commit:
Merge branch 'develop' of https://github.com/efim/habitrpg into Priority

Commit:
changed nemes of text strings on buttons and labels

efim committed:
changed task parameters affected by buttons, added recalculation of tasks weight

Commit:
added update of task.weight to onclick of buttons

efim commented:
https://github.com/HabitRPG/habitrpg-shared/pull/278 - corresponding habitrpg-shared pull-request

efim commented on views/shared/tasks/task.jade:
```diff
@@ -190,12 +190,22 @@ li(bindonce='list', bo-id='"task-"+task.id', ng-repeat='task in obj[list.type+"s
+              a.hint.difficulty-multiplier-help(href='https://trello.com/card/priority-multiplier/50e5d3684fe3a7266b0036d6/17', target='_blank', popover-title=env.t('difficultyHelpTitle'), popover-trigger='mouseenter', popover=env.t('difficultyHelpContent'))=env.t('difficulty')
+              //change to difficulty and also parameter of class to difficulty
+            .task-controls.tile-group.difficulty-multiplier
+              button.task-action-btn.tile(type='button', ng-class='{active: task.difficulty==0.5}', ng-click='task.challenge.id || ((task.difficulty=0.5) && (task.weight = task.priority * task.difficulty))')=env.t('easy')
```
Probably shouldn't have put task.weight update in ng-click, but didn't really know how to use $watch for class field, and didn't know if this would take a lot of resources.
(in all button's)

Commit:
button for trivial difficulty setting

crookedneighbor commented:
Here's the accompanying patch url for shared.

https://github.com/HabitRPG/habitrpg-shared/pull/278.patch

@efim said:

> view part for https://github.com/HabitRPG/habitrpg/pull/3887
> 
> creating difficulty settings for tasks (and etc.) under the priority settings in advanced optinons.
> Corresponding ticket - https://github.com/HabitRPG/habitrpg/issues/3556

@lemoness is the difficulty tool still desired?

lemoness commented:
Yes, definitely still desired! We want priority in addition to difficulty.
As I recall, this was fairly far along - what is the current status?

On Sun, Jan 18, 2015 at 5:51 PM, Blade Barringer notifications@github.com
wrote:

> Here's the accompanying patch url for shared.
> 
> https://github.com/HabitRPG/habitrpg-shared/pull/278.patch
> 
> @efim https://github.com/efim said:
> 
> view part for #3887 https://github.com/HabitRPG/habitrpg/pull/3887
> 
> creating difficulty settings for tasks (and etc.) under the priority
> settings in advanced optinons.
> Corresponding ticket - #3556
> https://github.com/HabitRPG/habitrpg/issues/3556
> 
> @lemoness https://github.com/lemoness is the difficulty tool still
> desired?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/3887#issuecomment-70437649.

efim commented:
@lemoness well as far as I can tell - it should work as desired if patch @crookedneighbor has provided  to be applied to according file, that was moved from "shared" repo.

This pull request is still waiting for someone to test it.

Will it be done more quickly if I spend some time setting up developing environment and incorporate patch into changed files in this request? Or if it is something tester can do easily it might be unnecessary.

=( So much time has passed and I am still waiting for this to get live.

lemoness commented:
Pinging @alys @lefnire @paglias as well.

On Wed, Mar 25, 2015 at 6:41 AM, efim notifications@github.com wrote:

> @lemoness https://github.com/lemoness well as far as I can tell - it
> should work as desired if patch @crookedneighbor
> https://github.com/crookedneighbor has provided to be applied to
> according file, that was moved from "shared" repo.
> 
> This pull request is still waiting for someone to test it.
> 
> Will it be done more quickly if I spend some time setting up developing
> environment and incorporate patch into changed files in this request? Or if
> it is something tester can do easily it might be unnecessary.
> 
> =( So much time has passed and I am still waiting for this to get live.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/3887#issuecomment-86029183.

Alys commented:
I'm about to start applying the patch and testing this. It changes some of the most basic, essential features of the site, so I might take a while to test. I don't want to rush through it.

Alys commented:
I've created a new branch for this: [task-difficulty-priority-importance-settings-2015-04](https://github.com/HabitRPG/habitrpg/tree/task-difficulty-priority-importance-settings-2015-04)
I've merged in efim's PR, including the patch from habitrpg-shared (thanks Blade!) All future PRs for this work should be made to that branch.

I've modified the priority and difficulty buttons to give them the new task page appearance (screenshot below).

The changes so far can be seen with this `compare` URL:
    https://github.com/HabitRPG/habitrpg/compare/task-difficulty-priority-importance-settings-2015-04?expand=1

I have so far done only the most basic testing. I'll do more once the points below are sorted out. Points are numbered for ease of reference in replies.
1. @lemoness, I'd like to suggest that instead of "priority" we use "importance". Importance has an unambiguous meaning. Priority can mean importance, or it can mean the order in which you intend to perform tasks. E.g., when the rubbish truck is coming down the street, taking out the rubbish has higher priority than taking your important medicine. I think we will have fewer questions from players about what the new buttons are for if we label them "importance". I am using "importance" below, but if we decide to stick with "priority", I'll edit this comment to match.
2. In the screenshot below, I've included the "Attributes" buttons, even though they are not affected by the PR; they are just included for comparison. Do we want to space the importance and difficulty buttons in the same way? I personally prefer the importance and difficulty buttons as they are, but others might disagree.
3. The new "trivial" button has been added to the importance setting, but shouldn't it be for difficulty? Or both for consistency and to avoid questions about why it's not on both? (I would REALLY like to avoid more eternally-repeated questions.)
4. @efim, task.weight is calculated in a few places (in the buttons and twice in index.coffee). Are all of those needed? 
5. Are we intending for difficulty and importance to have different effects? For example, when you complete a task, the difficulty alone controls how much XP and GP you earn; when you miss a Daily, the importance alone controls how much damage you take? Although that would make importance irrelevant for To-Dos. Or do we want both to affect earnings and damage, but difficulty to have more influence over earnings and importance to have more influence over damage? (Depending on our intentions here, task.weight might be unnecessary.)
6. We should decide on default difficulty and importance that a new task will have. On the current site, new tasks have an easy difficulty, which I think we should maintain for consistency. What do we want for the default importance? Low? Medium? _(Note to developers: "difficulty" here refers to what the users see in the edit form; on the backend it's stored as "priority". See the next point.)_
7. We should develop a strategy for converting all existing tasks. On the current site, tasks have a difficulty as far as the users are concerned, but in the backend this is stored as "priority" (importance). Thus with this PR as it is at the moment, all existing tasks would have an importance that "equals" their previous difficulty, and a new difficulty that is whatever default we choose. This is probably not what users will be expecting. I suggest we should run a migration script to convert all current "priority" settings to difficulty settings, and to add an importance setting. The importance setting that is added should be EITHER the default for new tasks, OR a value that is based on the task's current difficulty setting (so a medium difficulty task would be given a normal importance).
8. We should decide on the values for the levels of difficulty. With this PR as it is at the moment, they are 0.5, 1, and 2, whereas on the current site, the "priority" values are 1, 1.5, and 2. Given that the current "priority" is being presented to the users as the difficulty, I think that the new difficulty levels should have the same values.
9. Are we happy with the new importance setting having values of 1, 1.5, and 2 (and 0.1 for trivial)? I think that's good.
10. With the new code as it is at the moment, changes to difficulty aren't being saved so we'll need to fix that. That might be a bug introduced by the merging.
11. The help text for difficulty and importance should probably be adjusted to make it clearer, but how we do that might depend on the answers to some of the points above. 
    - This is the difficulty help text from the current site:
      "_This multiplies its point value. Use sparingly, rely instead on our organic value-adjustment algorithms. But some tasks are grossly more valuable (Write Thesis vs Floss Teeth). Click for more info._"
    - This is the difficulty help text from efim's PR:
      "_This multiplies its point value. Use sparingly, rely instead on our organic value-adjustment algorithms. But some tasks are much more hard while roughly equally valuable (Cooking Dinner for Whole Family vs Making Meal for Yourself). Click for more info._"
    - This is the importance/priority help text from efim's PR:
      "_This multiplies its point value. Some tasks are much more valuable while equally hard (Take Medicine vs Take Out Trash). Click for more info._"
    - Note that in the latter two texts, I changed "grossly" to "much" because it's probably better for ESL.
12. api.taskClasses will need to be adjusted to either change "priority" to difficulty, or to include both difficulty and importance, and also to include the trivial setting.
13. Existing tests will need to be modified. New tests might be needed.
14. A full search of the code should be done to ensure all places that contain "priority" have been adjusted as necessary. E.g., website/src/utils.js

_Note to Alys: Edit the [Data Display Tool](https://oldgods.net/habitrpg/habitrpg_user_data_display.html) when this goes live to correctly deal with difficulty and importance._

![screen shot 2015-04-12 at 7 12 16 pm](https://cloud.githubusercontent.com/assets/1495809/7105367/74a832d6-e158-11e4-8cb3-b1f9fef64d9b.png)

lemoness commented:
1. I like "Importance."
2. I think the buttons look fine!
3. I had also assumed that "Trivial" was for Difficulty, not Importance,
   and I believe that is what was intended on the original suggestion. I would
   be open to having it on both, but don't have strong feelings on the subject.
4. Importance was going to be more closely related to punishment, and
   Difficulty to reward, but you're correct that this makes Importance less
   relevant for To-Dos. This is open for discussion!
5. I think the default Importance should be "Normal."
6. Agreed!
7. Agreed!
8. Let's wait on text adjustments until we have everything else finalized
   so that we don't have to duplicate the work.

Thank you for all your work on this!

On Sun, Apr 12, 2015 at 4:12 AM, Alys notifications@github.com wrote:

> I've created a new branch for this:
> task-difficulty-priority-importance-settings-2015-04
> https://github.com/HabitRPG/habitrpg/tree/task-difficulty-priority-importance-settings-2015-04
> I've merged in efim's PR, including the patch from habitrpg-shared (thanks
> Blade!) All future PRs for this work should be made to that branch.
> 
> I've modified the priority and difficulty buttons to give them the new
> task page appearance (screenshot below).
> 
> The changes so far can be seen with this compare URL:
> 
> https://github.com/HabitRPG/habitrpg/compare/task-difficulty-priority-importance-settings-2015-04?expand=1
> 
> I have so far done only the most basic testing. I'll do more once the
> points below are sorted out. Points are numbered for ease of reference in
> replies.
> 
>    1.
> 
>    @lemoness https://github.com/lemoness, I'd like to suggest that
>    instead of "priority" we use "importance". Importance has an unambiguous
>    meaning. Priority can mean importance, or it can mean the order in which
>    you intend to perform tasks. E.g., when the rubbish truck is coming down
>    the street, taking out the rubbish has higher priority than taking your
>    important medicine. I think we will have fewer questions from players about
>    what the new buttons are for if we label them "importance". I am using
>    "importance" below, but if we decide to stick with "priority", I'll edit
>    this comment to match.
>    2.
> 
>    In the screenshot below, I've included the "Attributes" buttons, even
>    though they are not affected by the PR; they are just included for
>    comparison. Do we want to space the importance and difficulty buttons in
>    the same way? I personally prefer the importance and difficulty buttons as
>    they are, but others might disagree.
>    3.
> 
>    The new "trivial" button has been added to the importance setting, but
>    shouldn't it be for difficulty? Or both for consistency and to avoid
>    questions about why it's not on both? (I would REALLY like to avoid more
>    eternally-repeated questions.)
>    4.
> 
>    @efim https://github.com/efim, task.weight is calculated in a few
>    places (in the buttons and twice in index.coffee). Are all of those needed?
>    5.
> 
>    Are we intending for difficulty and importance to have different
>    effects? For example, when you complete a task, the difficulty alone
>    controls how much XP and GP you earn; when you miss a Daily, the importance
>    alone controls how much damage you take? Although that would make
>    importance irrelevant for To-Dos. Or do we want both to affect earnings and
>    damage, but difficulty to have more influence over earnings and importance
>    to have more influence over damage? (Depending on our intentions here,
>    task.weight might be unnecessary.)
>    6.
> 
>    We should decide on default difficulty and importance that a new task
>    will have. On the current site, new tasks have an easy difficulty, which I
>    think we should maintain for consistency. What do we want for the default
>    importance? Low? Medium? _(Note to developers: "difficulty" here
>    refers to what the users see in the edit form; on the backend it's stored
>    as "priority". See the next point.)_
>    7.
> 
>    We should develop a strategy for converting all existing tasks. On the
>    current site, tasks have a difficulty as far as the users are concerned,
>    but in the backend this is stored as "priority" (importance). Thus with
>    this PR as it is at the moment, all existing tasks would have an importance
>    that "equals" their previous difficulty, and a new difficulty that is
>    whatever default we choose. This is probably not what users will be
>    expecting. I suggest we should run a migration script to convert all
>    current "priority" settings to difficulty settings, and to add an
>    importance setting. The importance setting that is added should be EITHER
>    the default for new tasks, OR a value that is based on the task's current
>    difficulty setting (so a medium difficulty task would be given a normal
>    importance).
>    8.
> 
>    We should decide on the values for the levels of difficulty. With this
>    PR as it is at the moment, they are 0.5, 1, and 2, whereas on the current
>    site, the "priority" values are 1, 1.5, and 2. Given that the current
>    "priority" is being presented to the users as the difficulty, I think that
>    the new difficulty levels should have the same values.
>    9.
> 
>    Are we happy with the new importance setting having values of 1, 1.5,
>    and 2 (and 0.1 for trivial)? I think that's good.
>    10.
> 
>    With the new code as it is at the moment, changes to difficulty aren't
>    being saved so we'll need to fix that. That might be a bug introduced by
>    the merging.
>    11.
> 
>    The help text for difficulty and importance should probably be
>    adjusted to make it clearer, but how we do that might depend on the answers
>    to some of the points above.
>     - This is the difficulty help text from the current site: "_This
>       multiplies its point value. Use sparingly, rely instead on our organic
>       value-adjustment algorithms. But some tasks are grossly more valuable
>       (Write Thesis vs Floss Teeth). Click for more info._"
>       - This is the difficulty help text from efim's PR: "_This
>       multiplies its point value. Use sparingly, rely instead on our organic
>       value-adjustment algorithms. But some tasks are much more hard while
>       roughly equally valuable (Cooking Dinner for Whole Family vs Making Meal
>       for Yourself). Click for more info._"
>       - This is the importance/priority help text from efim's PR: "_This
>       multiplies its point value. Some tasks are much more valuable while equally
>       hard (Take Medicine vs Take Out Trash). Click for more info._"
>       - Note that in the latter two texts, I changed "grossly" to "much"
>       because it's probably better for ESL.
>     12.
> 
>    api.taskClasses will need to be adjusted to either change "priority"
>    to difficulty, or to include both difficulty and importance, and also to
>    include the trivial setting.
>    13.
> 
>    Existing tests will need to be modified. New tests might be needed.
>    14.
> 
>    A full search of the code should be done to ensure all places that
>    contain "priority" have been adjusted as necessary. E.g.,
>    website/src/utils.js
> 
> [image: screen shot 2015-04-12 at 7 12 16 pm]
> https://cloud.githubusercontent.com/assets/1495809/7105367/74a832d6-e158-11e4-8cb3-b1f9fef64d9b.png
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/3887#issuecomment-92033766.

efim commented:
@Alys 
1. I also thought at the time that 'importance' could be better fit. I tried to stick to more literal implementation of ticket though, because it was discussed at length at that point.
1. IIRC there was part of initialization code in coffee file and comment that it was needed for legacy issues, so i added initialization of weight as well, if it is still needed after almost a year I don't know.
   Weight is calculated in button actions because I've researched ways to bind variables (priority and difficulty to weight) in javascript and all those ways caused big load on user side.
2. That is completely up to you. I'd be happy with any result, as long as there is more flexibility in grading curve. I've implemented ticket literally and I think it is fine as is (i.e they have identical influence)
3. This could probably be solved by announcement only.
   Current 'Difficulty' is difficulty only in name, but essentially it is 'Importance'\'Priority' based on its description in bubble or wiki.
   IIRC this change shouldn't mess up peoples 'Difficulty' setting, but just rename it to 'Priority' since current difficulty is already stored in priority variable.
4. It was a while, not sure I can see where the problem is now.
5. This maybe was more recent addition to codebase than this change, I don't remember dealing with this.
6. This should definitely be included in testing process. I'm pretty certain I did thorough search at the time, but since then new dependencies could be introduced.

efim commented:
@Alys @lemoness 
Any problems with this I can help with? There was quite a long pause in discussion.

crookedneighbor commented:
@efim We're slowly building up our test suite as new prs come in. We'll need tests around these changes before we can merge in the PR. If you'd like to start writing tests, you can PR them into the branch Alys set up: https://github.com/HabitRPG/habitrpg/tree/task-difficulty-priority-importance-settings-2015-04

There may be some other small adjustments that need to be made, but the tests will take the most work, I think.

We'll also probably want to test this out on the beta site before merging it into the main one.

Alys commented:
Summary of the points above, using the same numbering scheme from my earlier comment and Lemoness's reply:
1. We will use "Importance" instead of "Priority".
2. Leave buttons as they are now (as shown in my screenshot).
3. A "Trivial" button should be available for both Difficulty and Importance. I think it would be good to have a different word for one of them, instead of Trivial for both, since the other Difficulty/Importance settings each use different words.
4. The weight calculations should be removed from anywhere they aren't necessary, which might end up being everywhere (see next point).
5. Importance will be more closely related to punishment, and Difficulty to reward, but there should be some overlap. I.e., Difficulty would have more influence over earnings, but Importance should have some effect. Importance would have more influence over damage, but Difficulty should have some too.
6. For new tasks created after this PR goes live, the default Difficulty and Importance should be Medium/Normal. I would love to have a preference setting for this though (I **really** like the current default of easy), but that could come later, if it's considered desirable by others too.
7. For existing tasks, when this change goes live, we will run a migration to set all values for Importance and Difficulty to the current values of Priority, and then delete the Priority settings (e.g., if an existing task has a Priority of Easy, it will be given a Difficulty of Easy and an Importance of Low). @efim, we can help with creation of the migration script (and indeed with anything else here, as you desire).
   - I believe we should also change the code to remove all mentions of `priority` and use `importance` and `difficulty` instead. I'm normally not a fan of adjusting variable names, but in this case there is potential for future programmer confusion (this is in no way the fault of this current change; we've always used `priority` in the back end for the thing that has been reported on the front end as "Difficulty").
8. The levels of Difficulty in this PR (0.5, 1, and 2) should be changed to those used on the current site for Priority (1, 1.5, and 2), for backwards consistency. The new Trivial setting can be 0.1.
9. The Importance values should be 0.1, 1, 1.5, and 2 to match.
10. With the new code as it is at the moment, changes to Difficulty aren't being saved so we'll need to fix that. That might be a bug introduced by the merging.
11. The help text for Difficulty and Importance should be adjusted but we will sort that out when everything else is finalised.
12. api.taskClasses will need to be adjusted to change Priority to Difficulty, to include Importance, and to include the Trivial settings.
13. Existing tests will need to be modified. New tests will be needed. @efim, we can assist with that as much as needed.

_Note to Alys: Edit the [Data Display Tool](https://oldgods.net/habitrpg/habitrpg_user_data_display.html) when this goes live to correctly deal with Difficulty and importance._

crookedneighbor commented:
@efim Since I haven't heard from you since April, I'm going to close this, but if you'd like to bring it up to date, I'll happily review it and help you write tests.

Alys commented:
@efim If you do want to keep working on this but don't have time at the moment for the full solution, we could start by adding just a "Trivial" button to the existing priority/difficulty setting.

Alys commented:
@efim I've copied your code for the Trivial difficulty button to https://github.com/HabitRPG/habitrpg/pull/5508 so that we can make that part of your work live soon.

efim commented:
Hello @Alys do you know whether anyone is working on that now?
I would like to resurrect this thing.

Also have you done tests for trivial button and where should I look for them?
[https://trello.com/c/4C8w1z5h/17-task-difficulty-settings-v2-priority-multiplier] is still quite popular and that is what I would really like to have for myself.

Alys commented:
I don't believe anyone is, and I'd like to have this too. I've been meaning to post here about it to see if you wanted to keep working on it.

There's a few tests related to priority in `test/api/todos.coffee` but I'm not sure how detailed they are.

I think we still need to sort out exactly how Difficulty(priority) and Importance will affect earnings and damage. My current thoughts are that Difficulty should affect earnings as it does now, but should stop having any effect on the punishments. Importance should take over the function of removing health and affecting the amount of damage to the party. This would mean that Importance would not be a setting for To-Dos. I'm also thinking that it should actually just be called "Damage" or something like that, to make it really clear what it does. But those are just my initial thoughts. :)
